test/zipkin/util.go

```diff
@@ -0,0 +1,135 @@
+
+	// ZipkinPort port exposed by the Zipkin Pod
+	ZipkinPort = 9411
+
+	// ZipkinTraceEndpoint port-forwarded zipkin endpoint
+	ZipkinTraceEndpoint = "http://localhost:9411/api/v2/trace/%s"
+
+	// ZipkinNamespace namespace where zipkin pod runs.
+	ZipkinNamespace = "istio-system"
+)
+
+// SetupZipkinTracing sets up zipkin tracing which involves a) Setting up port-forwarding from localhost to zipkin pod on the cluster (pid of the process doing Port-Forward is stored in an env variable).
+// b) enable AlwaysSample config for tracing.
+func SetupZipkinTracing(kubeClientset *kubernetes.Clientset) error {
+
+	logger := logging.GetContextLogger("SpoofUtil")
+	if !CheckZipkinPortAvailability() {
+		return errors.New("Zipkin port not available on the machine. Zipkin Setup failed")
+	}
+
+	zipkinPods, err := kubeClientset.CoreV1().Pods(ZipkinNamespace).List(metav1.ListOptions{LabelSelector: "app=zipkin"})
+
+	if err != nil {
+		return errors.New("Error retrieving Zipkin pod details. Zipkin Setup failed")
+	}
+
+	if len(zipkinPods.Items) == 0 {
+		return errors.New("No Zipkin Pod found. Ensure monitoring is switched on for your Knative Setup. Zipkin Setup failed")
+	}
+
+	portForwardCmd := exec.Command("kubectl", "port-forward", fmt.Sprintf("--namespace=%s", ZipkinNamespace), zipkinPods.Items[0].Name, fmt.Sprintf("%d:%d", ZipkinPort, ZipkinPort))
+	err = portForwardCmd.Start()
+
+	if err != nil {
+		return errors.New("Error starting kubectl port-forward command")
+
+	}
+
+	os.Setenv(ZipkinPortForwardPid, strconv.Itoa(portForwardCmd.Process.Pid))
+
+	// Applying AlwaysSample config to ensure we propagate zipkin header for every request made by this client.
+	trace.ApplyConfig(trace.Config{DefaultSampler: trace.AlwaysSample()})
+
+	logger.Infof("Successfully setup SpoofingClient for Zipkin Tracing")
+
+	return nil
+}
+
+// CleanupZipkinTracingSetup cleans up the Zipkin tracing setup on the machine. This involves killing the process performing port-forward.
+// Method expects the port-forwarding process to be stored in an environment variable.
+func CleanupZipkinTracingSetup() error {
+
+	logger := logging.GetContextLogger("SpoofUtil")
+	pid := os.Getenv(ZipkinPortForwardPid)
+
+	if len(pid) == 0 {
+		return errors.New("Zipking port-forward Process ID not found")
+	}
+
+	logger.Infof("Found Zipkin port-forward Process ID: %s", pid)
+
+	pidInt, err := strconv.Atoi(pid)
+
+	if err != nil {
+		return errors.New("Error converting Zipkin port-forward Process ID into an Int")
+	}
+
+	ps := os.Process{Pid: pidInt}
+	err = ps.Kill()
+
+	if err != nil {
+		return errors.New("Encoutered error killing process")
+	}
+
+	logger.Infof("Successfully killed Zipkin port-forward process")
+
+	return nil
+}
+
+// CheckZipkinPortAvailability checks to see if Zipkin Port is available on the machine.
+// If due to error it is not possible to determine if the port is available the method returns "false"
+func CheckZipkinPortAvailability() bool {
```

Comment:
Return an error instead of a `bool`.

Reply:
Is this more a Go pattern? coz in my opinion, error would not be appropriate as not having the zipkin port available is not essentially an error, rather the state of the system.

Reply:
You might have several reasons for the port not being available, thus you return the error explaining what was that instead of a generic "it didn't work" (line 130).

Reply:
Got your point and makes sense. So to make sure we also have information about the reason for port-unavailability I am returning both error and bool and logging the error where-ever it makes sense.

Reply:
This is unnecessary complexity. Return the error if something failed, return `nil` if no error.

Reply:
You win :)